src/lib/spinel/radio_spinel_impl.hpp

```diff
@@ -1301,7 +1301,7 @@ otError RadioSpinel<InterfaceType, ProcessContextType>::AddSrcMatchShortEntry(ui
     SuccessOrExit(error = Insert(SPINEL_PROP_MAC_SRC_MATCH_SHORT_ADDRESSES, SPINEL_DATATYPE_UINT16_S, aShortAddress));
 
 #if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0
-    assert(mSrcMatchShortEntryCount < OPENTHREAD_CONFIG_MLE_MAX_CHILDREN);
+    VerifyOrExit(mSrcMatchShortEntryCount < OPENTHREAD_CONFIG_MLE_MAX_CHILDREN);
```

Comment:
I am not sure changing to `VerifyOrExit()` would be good.

I am not myself familiar with `RCP_RESTORATION_MAX_COUNT` code (maybe @bukepo can provide more info)

In general, we can have at most `MLE_MAX_CHILDREN` children and assuming all are sleepy and would use indirect tx, the src match table should be limited by the `MLE_MAX_CHILDREN`? So how could we end up with more entries and trigger the assert?

Reply:
>  We have a multipan implementation with custom source match table and MLE_MAX_CHILDREN does not apply to all the stacks in question.

Do you inetnd to enable and use `CONFIG_RCP_RESTORATION` with multi-PAN RCP? 
If so, then probably need to update "RCP restoration" feature itself to be aware of multi-PAN RCP and save info for not just the Thread network but all PANs.

Reply:
FWIW, I believewWith the current implementation, OT will properly handle the situation where the source match table does not contain the entry as expected:
https://github.com/openthread/openthread/blob/73d3ea54491ae35896f46f9951f5bf52e46c34e9/src/core/mac/data_poll_handler.cpp#L152-L160.

Although it is less efficient than it should be. And looking at this now, we might want to log this situation if this is not expected, since this is effectively sweeping the inconsistency under the cover.

Reply:
@jwhui I agree and that is what I am expecting the stacks to do as well. In my opinion the source match table checks should be done by the RCP and the stack. SPINEL should be asserting if there are some inconsistencies.

Reply:
@abtink we are already using the MAX_RESTORATION count in our RCP and on the hosts. The only issue we see is handling the MAX_CHILDREN, which is handled differently for each stacks. As I mentioned above I just want to move the check from SPINEL to the stack/RCP code.

Reply:
My point is that if you use "RCP restoration" feature with multi-PAN RCP, it may not function correctly.

The idea behind "RCP restoration" feature is that `RadioSpinel` will save important radio info and ensure to restore them on RCP if it is reset unexpectedly. This includes a copy of "src match table" entries. If we use it with multi-PAN the "RCP restoration" feature, it should know to save the info for each PAN otherwise it will not work properly. We need a src match table per PAN (this may need to added in the multi-PAN PR itself).

Reply:
Alternative is to not use "RCP restoration" with multi-PAN RCP feature (which is what I would recommend).

Reply:
@abtink The idea is that each stack will have their own copy of `radio_spinel_impl` running and their own copies of source match tables. Introducing multipan awareness to that part of the code doesn't affect anything as both protocols are oblivious to each other and don't know about the others source match tables. Each protocol will maintain their own copies of restoration objects and will set them in RCP on reset using different IIDs.

Reply:
Than I wonder how the `assert` was triggered? 

Should each `radio_spinel_impl` have its own local src match table with its own `MAX_CHILDREN` entries (which we should never go beyond). 

Is the issue that the `MAX_CHILDREN` is not correctly set for some PAN? Maybe we should define a new config to specify the size of local table used for rcp restoration?

Reply:
Not really, so if a protocol supports 802.15.4 association (like zigbee), all joining devices join using indirect messaging. So lets say a router is joining, it will also claim a source match entry, even though it is not technically a child. So to overcome this we have a big enough source match table on RCP but the stack does not define a big `MAX_CHILDREN` causing this issue.

Reply:
> That is why I am suggesting the stacks to validate the src match table instead of the SPINEL doing this.

`RadioSpinel` is not really validating the src match table.

Thread stack already validates and manages src match table:
- It detects when table is full when `otPlatRadioAddSrcMatchEntry()` function return `error`.
- It then disables src match table and enables it back when we can fit all entries back in the table.
- It keeps track of which `Child` entries are not yet updated/added in `SrcMatchTable`
(these are all implemented in `SrcMatchTable` and `Child` classes)

 This `assert` in `RadioSpinel` is ensuring its own configuration is correct, so that it can function correctly.
- `RadioSpinel` for "RCP restoration" feature (and only when this feature is enabled) keeps a copy of src match table so it can restore it on RCP (if it resets).
- The `asssert `in `RadioSpinel` is to make sure the local table size in `RadioSpinel` is larger enough.
  - In `RadioSpinel` (and/or OT core) we don't really know how many entries radio/RCP supports, so instead it assumes `MAX_CHILDREN` which it knows is the max Thread stack host will use.
  
My concern with removing the `assert` (or changing it to `VerifyOrExit`) is that it can actually mask an incorrect configuration. 

------

> Not really, so if a protocol supports 802.15.4 association (like zigbee), all joining devices join using indirect messaging. So lets say a router is joining, it will also claim a source match entry, even though it is not technically a child. So to overcome this we have a big enough source match table on RCP but the stack does not define a big MAX_CHILDREN causing this issue.

This is not clear to me. My understating from your earlier answer is that each protocol will be using its own process and its own `RadioSpinel` instance:
- The `RadioSpinel` used with Thread stack would at most have `MAX_CHILDREN` indirect tx.
- If you use a `RadioSpinel` with another stack (say Zigbee), it can define its own `MAX_CHIDLREN` config for what it uses, any reason this would not work?

Reply:
@abtink I understand your first comment, but the reasoning for this change ties in with the second point.

About the second comment yes:

- Each protocol does use its own own SPINEL.
- Each protocol do define its own MAX_CHILDREN.
- But MAX_CHILDREN indicates the number of children supported by a stack.
- For thread this is also equal to src match table size, because child devices are the only ones that will do indirect tx.
- For Zigbee this is not true. Even if Zigbee defines its own MAX_CHILDREN, devices other than child devices (routers), also use the indirect queue for joining. This means that the assert will trigger even if RCP has capability  to support more children and wont report any errors, because in zigbee MAX_CHILDREN != src match table size.

One way to deal with it is to increase the MAX_CHILDREN to include non-child devices as well (which will basically be the network size in star topology), but it is not really a good idea, as other entries are just empty once the network is established. The other option is to allow RCP to accomodate the bigger src match table when needed, and not really be concerned if the spinel copy of src match table is not up-to-date.

Reply:
My suggestion is different:
- When building `RadioSpinel` for use with ZigBee (or any other protocol), we can still use the `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN` config to specify the size of the src match table used by `RadioSpinel` and we can set it to whatever make sense for that stack. 
   - The `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN` config is OpenThread (and Thread protocol) specific.
   - Other stacks probably have their own configs to define max children supported by the stack (they should not use `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN` as a way to define how many children they support).

---

A second suggestion (which I mentioned earlier) is to define a new spinel specific config for src match table size of `RadioSpinel`. This way we avoid any confusion with its association with `MAX_CHILDREN`:
- We can define a new spinel specific config `OPENTHREAD_SPINEL_CONFIG_MAX_SRC_MATCH_ENTRIES` in `openthread-spinel-config.h` (defines size of the  local src match table used by `RadioSpinel` when `RCP_RESTORATION` feature is enabled and used).
- We set the default value for this config to be `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN` (so it works correctly under OT and Thread stack).
- Other protocols can define whatevery value they want for size of the table using this config.


---

> The other option is to allow RCP to accomodate the bigger src match table when needed, and not really be concerned if the spinel copy of src match table is not up-to-date.

This is about host `RadioSpinel` and whether we want RCP restoration to function correctly or not. 
- If you want to use RCP restoration feature of `RadioSpinel` when used with other protocols/stacks, then we need to make sure it can keep a local copy of src match table that is the same as what is set on RCP. 
- That's why I asked earlier whether you intend to use the "RCP restoration feature" with other stacks (under multi-PAN).

Reply:
@hemanth-silabs @romacdon , any further thoughts on this PR?